src/live_restore/live_restore_server.c

```diff
@@ -0,0 +1,265 @@
+ *  All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+#include "live_restore_private.h"
+
+/*
+ * __live_restore_worker_check --
+ *     Thread groups cannot exist without a check function but in our case we don't use it due to it
+ *     not meshing well with how we terminate threads. Given that, this function simply returns
+ *     true.
+ */
+static bool
+__live_restore_worker_check(WT_SESSION_IMPL *session)
+{
+    WT_UNUSED(session);
+    return (true);
+}
+
+/*
+ * __live_restore_work_queue_drain --
+ *     Drain the work queue of any remaining items. This function does not need to take the queue
+ *     lock but we do anyway because the semantic is if you touch the queue after it is initialized,
+ *     you hold the lock.
+ */
+static void
+__live_restore_work_queue_drain(WT_SESSION_IMPL *session)
+{
+    WT_LIVE_RESTORE_SERVER *server = &S2C(session)->live_restore_server;
+    __wt_spin_lock(session, &server->queue_lock);
+    if (!TAILQ_EMPTY(&server->work_queue)) {
+        WT_LIVE_RESTORE_WORK_ITEM *work_item = NULL, *work_item_tmp = NULL;
+        TAILQ_FOREACH_SAFE(work_item, &server->work_queue, q, work_item_tmp)
+        {
+            TAILQ_REMOVE(&server->work_queue, work_item, q);
+            __wt_free(session, work_item->uri);
+            __wt_free(session, work_item);
+        }
+    }
+    WT_ASSERT(session, TAILQ_EMPTY(&server->work_queue));
+    __wt_spin_unlock(session, &server->queue_lock);
+}
+
+/*
+ * __live_restore_worker_run --
+ *     Entry function for a live restore thread. This is called repeatedly from the thread group
+ *     code so it does not need to loop itself.
+ */
+static int
+__live_restore_worker_run(WT_SESSION_IMPL *session, WT_THREAD *ctx)
+{
+    WT_DECL_RET;
+    WT_UNUSED(ctx);
+    WT_LIVE_RESTORE_SERVER *server = &S2C(session)->live_restore_server;
+
+    __wt_spin_lock(session, &server->queue_lock);
+    if (TAILQ_EMPTY(&server->work_queue)) {
+        /*
+         * Because we depend on stopping threads to avoid decrementing the counter indefinitely this
+         * section of code will both decrement the counter and terminate the thread. This is an
+         * unusual usage of thread groups but avoids refactoring a LOT of other code for the time
+         * being. The obvious fix would be to pass a WT_THREAD to the check function and then clear
+         * the run flag, however this then modifies the definition of __wt_cond_wait_signal which is
+         * used _everywhere_.
+         */
+        server->threads_working--;
+        if (server->threads_working == 0) {
+            /* All threads have finished their work, signal that source can be dropped. */
+            WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_COMPLETE);
+        }
+        /* Kill ourselves. */
+        F_CLR(ctx, WT_THREAD_RUN);
+        __wt_verbose_debug2(session, WT_VERB_FILEOPS, "Live restore worker terminating%s", "!");
```

Comment:
There are reasons that the verbose functions require a format and at least one arg. KeithB did the original work and would have "fixed" it if it was easy and feasible. I don't remember the details of the reasons. In any case, the consistent way to use it in this type of situation is like examples in `conn_close`. The format is just a string and the arg is your full desired string. This applies to all forms of verbose functions:
`__wt_verbose_info(session, WT_VERB_RECOVERY_PROGRESS, "%s", "closing WiredTiger library.");`

Reply:
03fa92875e32aea444e8e494abd7fa49cb72c255